### PR TITLE
feat(rule_engin): add two size-related functions

### PR DESCRIPTION
**Pull request by lafirest**
Fixes EMQX-13603

Release version: v/e5.8.4

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_rule_engine/src/emqx_rule_funcs.erl [hunk 1]
```diff
@@ -785,6 +787,15 @@ is_map(_) -> false.
...
 is_array(_) -> false.
 
+is_empty([]) ->
+    true;
+is_empty(<<>>) ->
+    true;
+is_empty(List) when is_list(List) ->
+    false;
+is_empty(Map) ->
+    ?MODULE:map_size(Map) == 0.
```
Comment (ieQu1) [inline apps/emqx_rule_engine/src/emqx_rule_funcs.erl:797]: Why use remote call (?MODULE:...)?

File: apps/emqx_rule_engine/src/emqx_rule_funcs.erl [hunk 2]
```diff
@@ -785,6 +787,15 @@ is_map(_) -> false.
 is_array(T) when is_list(T) -> true;
 is_array(_) -> false.
 
+is_empty([]) ->
+    true;
+is_empty(<<>>) ->
+    true;
```
[the remark this answers is not in the document]
Comment (lafirest) [reply apps/emqx_rule_engine/src/emqx_rule_funcs.erl:793]: No, the function is used for maps or arrays, non-empty binaries will be treated as encoded JSON

File: apps/emqx_rule_engine/src/emqx_rule_funcs.erl (hunk 1 above)
Comment (lafirest) [reply apps/emqx_rule_engine/src/emqx_rule_funcs.erl:797]: map_size is a built-in function of Erlang. Here we need to call another function of the same name in this module.

**Review by ieQu1 — APPROVED** [review]

**Review by savonarola — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-5386)

**Pull request merged**